pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,439 @@
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	foundInstance, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	switch instanceStatus(foundInstance.Status) {
+	case stoppedStatus:
+		startRequest := ecs.CreateStartInstanceRequest()
+		startRequest.InstanceId = foundInstance.InstanceId
+
+		_, err = client.StartInstance(startRequest)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("instance %v is in a stopped state", foundInstance.InstanceId)
+	case runningStatus:
+		if len(foundInstance.PublicIpAddress.IpAddress) == 0 {
+			ipAddress := ecs.CreateAllocatePublicIpAddressRequest()
+			ipAddress.InstanceId = foundInstance.InstanceId
+
+			_, err = client.AllocatePublicIpAddress(ipAddress)
+			if err != nil {
+				return nil, err
+			}
+			if err = data.Update(machine, func(updatedMachine *v1alpha1.Machine) {
+				if !kuberneteshelper.HasFinalizer(updatedMachine, finalizerPublicIP) {
+					updatedMachine.Finalizers = append(updatedMachine.Finalizers, finalizerPublicIP)
+				}
+			}); err != nil {
+				return nil, err
+			}
+		}
+		return &alibabaInstance{instance: foundInstance}, nil
+	}
+
+	return nil, fmt.Errorf("instance %v is not ready", foundInstance.InstanceId)
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Labels == nil {
+		c.Labels = map[string]string{}
+	}
+
+	c.Labels[machineUIDTag] = string(machine.UID)
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = p.getImageIDForOS(machine.Spec, pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	_, err = client.CreateInstance(createInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create instance at Alibaba cloud: %v", err)
```

Comment:
The finalizer is needed for the instance, not for the IP and must be added right here, once the createInstance request was done without an error. You do the `getIntance` call afterwards and that one may fail. We must make sure we add a finalizer as soon as we created something.

Reply:
addressed here: https://github.com/kubermatic/machine-controller/pull/643/files#diff-e96854fadd3b15014f29f87ac940a01bR234-R247